File: apps/windows/src/Shared/WebhookSender.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

namespace VoiceVector.Shared
{
    /// <summary>Folder webhook delivery — payload per docs/webhook-payload.md.</summary>
    public static class WebhookSender
    {
        public static string BuildPayload(Entry entry, string app = "voicevector-windows")
        {
            return Json.Write(new Dictionary<string, object>
            {
                { "app", app },
                { "id", entry.Id },
                { "folder", entry.Folder },
                { "date", Library.RenderDate(entry.Date) },
                { "duration", entry.Duration },
                { "raw", entry.Raw },
                { "cleaned", entry.Cleaned },
                { "stt", entry.SttLabel },
                { "cleanup", entry.CleanupLabel },
            });
        }

        public static async Task SendAsync(Entry entry, string audioPath, WebhookConfig config)
        {
            Uri url;
            if (!config.Enabled || !Uri.TryCreate(config.Url, UriKind.Absolute, out url)) return;

            for (int attempt = 1; attempt <= 3; attempt++)
            {
                try
                {
                    using (var request = new HttpRequestMessage(HttpMethod.Post, url))
                    {
                        if (config.IncludeAudio && File.Exists(audioPath))
                        {
                            var form = new MultipartFormDataContent();
                            form.Add(new StringContent(BuildPayload(entry)), "payload");
                            var audio = new ByteArrayContent(File.ReadAllBytes(audioPath));
                            audio.Headers.ContentType = new MediaTypeHeaderValue("audio/wav");
                            form.Add(audio, "audio", entry.AudioFilename);
                            request.Content = form;
                        }
                        else
                        {
                            request.Content = new StringContent(BuildPayload(entry),
                                                                Encoding.UTF8, "application/json");
                        }
                        using (var response = await ProviderClient.Http.SendAsync(request).ConfigureAwait(false))
                        {
                            response.EnsureSuccessStatusCode();
                        }
                        return;
                    }
                }
                catch (Exception e)
                {
                    Log.Error("Webhook attempt " + attempt + " for " + entry.Id + " failed: " + e.Message);
                    if (attempt < 3)
                        await Task.Delay(TimeSpan.FromSeconds(2 * attempt)).ConfigureAwait(false);
                }
            }
        }
    }
}

File: apps/macos/Sources/VoiceVector/Webhook/WebhookSender.swift
import Foundation

/// Forwards a finished dictation to the folder's webhook. JSON body by default;
/// multipart with the WAV attached when `includeAudio` is on. Two retries with
/// backoff; failures are logged, never block the pipeline.
enum WebhookSender {
    static func send(entry: Entry, audioURL: URL, config: WebhookConfig) async {
        guard config.enabled, let url = URL(string: config.url), !config.url.isEmpty else { return }

        var request: URLRequest
        let metadata: [String: Any] = [
            "app": "voicevector-macos",
            "id": entry.id,
            "folder": entry.folder,
            "date": Library.dateFormatter.string(from: entry.date),
            "duration": entry.duration,
            "raw": entry.raw,
            "cleaned": entry.cleaned,
            "stt": entry.sttLabel,
            "cleanup": entry.cleanupLabel,
        ]

        do {
            if config.includeAudio, let audio = try? Data(contentsOf: audioURL) {
                var form = Multipart()
                let json = try JSONSerialization.data(withJSONObject: metadata)
                form.addField(name: "payload", value: String(data: json, encoding: .utf8) ?? "{}")
                form.addFile(name: "audio", filename: entry.audioFilename,
                             contentType: "audio/wav", data: audio)
                request = HTTP.request(url, method: "POST", headers: [:],
                                       body: form.encoded(), contentType: form.contentType)
            } else {
                let json = try JSONSerialization.data(withJSONObject: metadata)
                request = HTTP.request(url, method: "POST", headers: [:],
                                       body: json, contentType: "application/json")
            }
        } catch {
            Log.error("Webhook payload build failed: \(error.localizedDescription)")
            return
        }

        for attempt in 1...3 {
            do {
                _ = try await HTTP.send(request)
                Log.info("Webhook delivered for \(entry.id)")
                return
            } catch {
                Log.error("Webhook attempt \(attempt) for \(entry.id) failed: \(error.localizedDescription)")
                if attempt < 3 {
                    try? await Task.sleep(nanoseconds: UInt64(attempt) * 2_000_000_000)
                }
            }
        }
    }
}
